common/rpc/encryption/localStoreCertProvider.go

```diff
@@ -95,101 +128,142 @@ func (s *localStoreCertProvider) FetchServerRootCAsForClient(isWorker bool) (*x5
 	if clientSettings == nil {
 		return nil, nil
 	}
-	rootCAFiles := clientSettings.RootCAFiles
-	rootCAData := clientSettings.RootCAData
+	certs, err := s.getCerts()
+	if err != nil {
+		return nil, err
+	}
+	if certs == nil {
+		return nil, nil
+	}
 
-	var cached **x509.CertPool
-	var certs *[]*x509.Certificate
 	if isWorker {
-		cached = &s.serverCAsWorkerPool
-		certs = &s.serverCACertsWorker
-	} else {
-		cached = &s.serverCAPool
-		certs = &s.serverCACerts
+		return certs.serverCAsWorkerPool, nil
 	}
 
-	return s.fetchCAs(
-		rootCAFiles,
-		rootCAData,
-		cached,
-		certs,
-		"cannot specify both rootCAFiles and rootCAData properties")
+	return certs.serverCAPool, nil
 }
 
-func (s *localStoreCertProvider) fetchCAs(
-	files []string,
-	data []string,
-	cachedCertPool **x509.CertPool,
-	certs *[]*x509.Certificate,
-	duplicateErrorMessage string) (*x509.CertPool, error) {
-	if len(files) == 0 && len(data) == 0 {
+func (s *localStoreCertProvider) FetchClientCertificate(isWorker bool) (*tls.Certificate, error) {
+
+	if s.tlsSettings == nil && s.workerTLSSettings == nil {
 		return nil, nil
 	}
+	certs, err := s.getCerts()
+	if err != nil {
+		return nil, err
+	}
+	if isWorker {
+		return certs.workerCert, nil
+	}
+	return certs.serverCert, nil
+}
+
+func (s *localStoreCertProvider) GetExpiringCerts(timeWindow time.Duration,
+) (CertExpirationMap, CertExpirationMap, error) {
+
+	expiring := make(CertExpirationMap)
+	expired := make(CertExpirationMap)
+	when := time.Now().UTC().Add(timeWindow)
+
+	certs, err := s.getCerts()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	checkError := checkTLSCertForExpiration(s.certs.serverCert, when, expiring, expired)
+	err = appendError(err, checkError)
+	checkError = checkTLSCertForExpiration(s.certs.clientCert, when, expiring, expired)
+	err = appendError(err, checkError)
+	checkError = checkTLSCertForExpiration(s.certs.workerCert, when, expiring, expired)
+	err = appendError(err, checkError)
+
+	checkCertsForExpiration(certs.clientCACerts, when, expiring, expired)
+	checkCertsForExpiration(certs.serverCACerts, when, expiring, expired)
+	checkCertsForExpiration(certs.serverCACertsWorker, when, expiring, expired)
+
+	return expiring, expired, err
+}
+
+func (s *localStoreCertProvider) getCerts() (*certCache, error) {
 
 	s.RLock()
```

Comment:
alternative - rather than having a goroutine that runs in the background, you can have a lastRefreshedAt and just do the refresh dynamically if the lastRefreshedAt + refreshInterval < currentTime.

Downside is that you increase locking time for this piece of code (because in the current implementation, you load and then lock whereas with the alternative, you lock and then load)

Reply:
Yeah, I'm uncomfortable blocking on the hot path of getting a cert, which is now performed on every connection attempt. Background refreshing is better I think.